File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.swift
import Foundation

func solution(_ n: Int, _ costs: [[Int]]) -> Int {
    // Kruskal + Union-Find(DSU)

    var parent = Array(0..<n)
    var rank = Array(repeating: 0, count: n)

    func find(_ x: Int) -> Int {
        if parent[x] == x { return x }
        parent[x] = find(parent[x])   // path compression
        return parent[x]
    }

    func union(_ a: Int, _ b: Int) -> Bool {
        var ra = find(a)
        var rb = find(b)
        if ra == rb { return false }

        if rank[ra] < rank[rb] { swap(&ra, &rb) }
        parent[rb] = ra
        if rank[ra] == rank[rb] { rank[ra] += 1 }
        return true
    }

    let sortedCosts = costs.sorted { $0[2] < $1[2] }

    var answer = 0
    var picked = 0

    for e in sortedCosts {
        let u = e[0], v = e[1], w = e[2]
        if union(u, v) {
            answer += w
            picked += 1
            if picked == n - 1 { break }
        }
    }

    return answer
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
def solution(n, costs):
    # Kruskal + Union-Find(DSU)

    parent = list(range(n))
    rank = [0] * n

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]  # path compression (halving)
            x = parent[x]
        return x

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra == rb:
            return False
        if rank[ra] < rank[rb]:
            ra, rb = rb, ra
        parent[rb] = ra
        if rank[ra] == rank[rb]:
            rank[ra] += 1
        return True

    costs.sort(key=lambda x: x[2])

    answer = 0
    picked = 0
    for u, v, w in costs:
        if union(u, v):
            answer += w
            picked += 1
            if picked == n - 1:
                break

    return answer